In [1]:
import pandas as pd
df = pd.read_csv("/content/sample_data/california_housing_train.csv")
df.head()

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
0,-114.31,34.19,15.0,5612.0,1283.0,1015.0,472.0,1.4936,66900.0
1,-114.47,34.40,19.0,7650.0,1901.0,1129.0,463.0,1.8200,80100.0
2,-114.56,33.69,17.0,720.0,174.0,333.0,117.0,1.6509,85700.0
3,-114.57,33.64,14.0,1501.0,337.0,515.0,226.0,3.1917,73400.0
4,-114.57,33.57,20.0,1454.0,326.0,624.0,262.0,1.9250,65500.0


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error


# Task
Import the necessary libraries, separate the dataset into features (X) and target (y), split the data into training and testing sets, train a linear regression model on the training data, and then make predictions on the test data.

## Separate features and target

### Subtask:
Define the features (X) and the target variable (y) for the model.


**Reasoning**:
Define the features (X) and the target variable (y) by dropping and selecting the appropriate columns from the DataFrame.



In [2]:
X = df.drop('median_house_value', axis=1)
y = df['median_house_value']

## Split data

### Subtask:
Split the data into training and testing sets.


**Reasoning**:
Split the features and target into training and testing sets using train_test_split with the specified test size and random state.



In [3]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

NameError: name 'train_test_split' is not defined

**Reasoning**:
The previous command failed because `train_test_split` was not recognized. I need to ensure the necessary library is imported before using the function. The previous code block already imported the necessary libraries. I will regenerate the code block to ensure all necessary imports are included.



In [4]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (13600, 8)
X_test shape: (3400, 8)
y_train shape: (13600,)
y_test shape: (3400,)


## Train the model

### Subtask:
Train a linear regression model using the training data.


**Reasoning**:
Instantiate and train a linear regression model using the training data.



In [5]:
model = LinearRegression()
model.fit(X_train, y_train)

NameError: name 'LinearRegression' is not defined

**Reasoning**:
The LinearRegression class was not imported in the previous code block, causing a NameError. It needs to be imported before instantiation.



In [6]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

## Test the model

### Subtask:
Make predictions on the test data and evaluate the model's performance.


**Reasoning**:
Use the trained model to make predictions on the test data and calculate the Mean Squared Error.



In [7]:
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse}")

NameError: name 'mean_squared_error' is not defined

**Reasoning**:
The previous attempt failed because `mean_squared_error` was not imported. Import the function and re-run the prediction and evaluation steps.



In [8]:
from sklearn.metrics import mean_squared_error

y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse}")

Mean Squared Error: 4634658406.223264


## Summary:

### Data Analysis Key Findings

*   The dataset was successfully split into training (80%) and testing (20%) sets.
*   A Linear Regression model was successfully trained on the training data.
*   Predictions were made on the test data using the trained model.
*   The Mean Squared Error (MSE) of the model's performance on the test data is 4,634,658,406.22.

### Insights or Next Steps

*   The MSE value is quite large, indicating that the model's predictions have a significant average squared error.
*   Further analysis is needed to understand the reasons for the high MSE. This could involve exploring feature engineering, different model types, or hyperparameter tuning.


# Task
Import necessary libraries, load the dataset, separate features and target, split the data into training and testing sets, train a linear regression model, load the test data from "/content/sample_data/california_housing_test.csv", make predictions on the test data, and evaluate the model using accuracy score.

## Test the model

### Subtask:
Load the separate test data from `/content/sample_data/california_housing_test.csv`, make predictions on this new test data, and evaluate the model's performance using the accuracy score.


**Reasoning**:
Load the new test data, separate features and target, make predictions, and calculate the Mean Squared Error.



In [9]:
df_new_test = pd.read_csv("/content/sample_data/california_housing_test.csv")
X_new_test = df_new_test.drop('median_house_value', axis=1)
y_new_test = df_new_test['median_house_value']
y_new_pred = model.predict(X_new_test)
mse_new = mean_squared_error(y_new_test, y_new_pred)
print(f"Mean Squared Error on new test data: {mse_new}")

Mean Squared Error on new test data: 4864606035.235156


## Summary:

### Data Analysis Key Findings

*   The Mean Squared Error (MSE) on the new test data was calculated to be 4864606035.235156.
*   Accuracy score was not used as it is not an appropriate metric for regression tasks.

### Insights or Next Steps

*   The high MSE suggests that the linear regression model may not be performing well on this new test data.
*   Further analysis should be conducted to understand the reasons for the high MSE, potentially involving feature engineering, exploring different models, or investigating the characteristics of the new test data compared to the training data.
